crates/accelerate/src/sabre_swap/mod.rs

```diff
@@ -621,103 +632,102 @@ fn route_reachable_nodes<F>(
 }
 
 /// Walk through the swaps in the given vector, undoing them on the layout and removing them.
-fn undo_swaps(swaps: &mut Vec<[VirtualQubit; 2]>, layout: &mut NLayout) {
-    swaps
-        .drain(..)
-        .rev()
-        .for_each(|swap| layout.swap_virtual(swap[0], swap[1]));
+fn undo_swaps(
+    swaps: &mut Vec<[PhysicalQubit; 2]>,
+    front_layer: &mut FrontLayer,
+    extended_set: &mut ExtendedSet,
+    layout: &mut NLayout,
+) {
+    swaps.drain(..).rev().for_each(|swap| {
+        front_layer.apply_swap(swap);
+        extended_set.apply_swap(swap);
+        layout.swap_physical(swap[0], swap[1]);
+    });
 }
 
 /// Find the node index and its associated virtual qubits that is currently the closest to being
 /// routable in terms of number of swaps.
-fn closest_operation(
-    front_layer: &FrontLayer,
-    layout: &NLayout,
-    dist: &ArrayView2<f64>,
-) -> (NodeIndex, [VirtualQubit; 2]) {
-    let (&node, qubits) = front_layer
+fn closest_operation<'a>(
+    front_layer: &'a FrontLayer,
+    dist: &'_ ArrayView2<f64>,
+) -> (&'a NodeIndex, &'a [PhysicalQubit; 2]) {
+    front_layer
         .iter()
-        .map(|(node, qubits)| (node, [qubits[0].to_phys(layout), qubits[1].to_phys(layout)]))
         .min_by(|(_, qubits_a), (_, qubits_b)| {
             dist[[qubits_a[0].index(), qubits_a[1].index()]]
                 .partial_cmp(&dist[[qubits_b[0].index(), qubits_b[1].index()]])
                 .unwrap_or(Ordering::Equal)
         })
-        .unwrap();
-    (node, [qubits[0].to_virt(layout), qubits[1].to_virt(layout)])
+        .unwrap()
 }
 
 /// Add the minimal set of swaps to the `swaps` vector that bring the two `qubits` together so that
 /// a 2q gate on them could be routed.
 fn swaps_to_route(
-    swaps: &mut Vec<[VirtualQubit; 2]>,
-    qubits: &[VirtualQubit; 2],
-    layout: &NLayout,
+    swaps: &mut Vec<[PhysicalQubit; 2]>,
+    qubits: &[PhysicalQubit; 2],
     coupling_graph: &DiGraph<(), ()>,
 ) {
     let mut shortest_paths: DictMap<NodeIndex, Vec<NodeIndex>> = DictMap::new();
```

Comment:
Completely unrelated to anything in this PR and not something we should consider for this but instead a follow up. But reading through the code now I'm wondering if we should do something like:

```suggestion
    let source_index = NodeIndex::new(qubits[0].index());
    let target_index = NodeIndex::new(qubits[1].index()));
    let mut shortest_paths: DictMap<NodeIndex, Vec<NodeIndex>> = [(node_index, Vec::with_capacity(distance[[source_index, target_index]])].iter().collect();
```

which definitely won't compile because I'm sure I made a typo or missed some typing, but it's the basic idea.

Reply:
Actually I just looked up the dijkstra code in rustworkx. This won't work because of: https://github.com/Qiskit/rustworkx/blob/2a2a18383ee877c7bc475ef15030beaea6d03520/rustworkx-core/src/shortest_path/dijkstra.rs#L120-L122 the rest of the code will handle this correctly. I can push a PR to rustworkx to enable this optimization so it detects if the path list is already allocated and just clears it and pushes the start node at the beginning. But until that is in rustworkx we can't do this. Not that reducing the allocations here will have a noticeable impact on runtime performance as this is only triggered in an edge case.